src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -957,37 +957,14 @@ public void PersistOnFlush(object obj)
 
 		public override string BestGuessEntityName(object entity)
 		{
-			using (BeginContext())
-			{
-				if (entity.IsProxy())
-				{
-					INHibernateProxy proxy = entity as INHibernateProxy;
-					ILazyInitializer initializer = proxy.HibernateLazyInitializer;
-
-					// it is possible for this method to be called during flush processing,
-					// so make certain that we do not accidentally initialize an uninitialized proxy
-					if (initializer.IsUninitialized)
-					{
-						return initializer.PersistentClass.FullName;
-					}
-					entity = initializer.GetImplementation();
```

Comment:
The entity name may be different from the proxy entity name. Getting the entity name can be more precise than getting the entity name known by the proxy.
This happens in case of mapped inheritance, when we have a proxy of the base class. When the proxy is initialized, the entity name will be the actual concrete implementation entity name, while the entity name known by the proxy will only be the one of the base class.
So the change here is a possible breaking change: with the change, entity-name guessed from proxies on base classes will always be the base class entity name. With current code it would be the concrete class entity name for initialized proxies.

Reply:
And now old code makes sense :) Fixed and commented.